### PR TITLE
Zero length string now handled for the Depth property

### LINKED ISSUE #554: XML_Toolkit: Handle nulls for Depth property
#### Description:

Markup objects with null <Depth /> property currently not handled correctly. Getting this error in Grasshopper:
image

image

Suggested fix is to use a nullable Double

#### Test file(s):

Example XML: https://burohappold.sharepoint.com/:u:/r/sites/049211/Shared%20Documents/F42%20Sustainability%20%26%20Physics/Stg4/SAP/B10/BP-RJA-10-XX-DRG-A-20-1001_combined.xml?csf=1&web=1&e=FbdTWm

Test Script:
https://burohappold.sharepoint.com/:u:/r/sites/Direction/designtechnology/RandDWishlist/00205_Environments%20%26%20MEP/SAP%20workflow/220524%20Dwellings%20from%20Bluebeam.gh?csf=1&web=1&e=3SPnIp

### PR DESCRIPTION
**Pull request by jamesramsden-bh**
Closes #554 

 

 ### Test files

Try different values of 'depth' in the XML file. Verify the value of 'Depth' in the exploded markup. 
[test.zip](https://github.com/BHoM/XML_Toolkit/files/8762243/test.zip)

 ### Changelog

 ### Additional comments

### CODE REVIEW

**Review by rboulton-BH — DISMISSED** [review]
I've tested the updated code using the sample dwelling and script which is referenced in the original issue. Looks like it's working well now with these changes

**Comment by FraserGreenroyd** [comment]
@BHoMBot check compliance

**Comment by jamesramsden-bh** [comment]
@BHoMBot check compliance

**Comment by jamesramsden-bh** [comment]
@BHoMBot check compliance

**Comment by jamesramsden-bh** [comment]
@BHoMBot check required

**Review by EKAdebo — DISMISSED** [review]
Looks like it's working, the data from XML files is coming through as expected.

**Comment by jamesramsden-bh** [comment]
@BHoMBot check ready-to-merge

**Comment by jamesramsden-bh** [comment]
@BHoMBot check ready-to-merge

**Review by FraserGreenroyd — CHANGES_REQUESTED** [review]
As discussed on a call with @jamesramsden-bh and @CKBoulter just now, agreement to add a `Query` method to query the double depth to support numerical workflows.

```
public static double Depth(this Markup markup)
{
    try
    {
        return System.Convert.ToDouble(markup.XMLDepth);
    }
    catch(Exception e)
    {
        BH.Engine.Base.Compute.RecordError("Could not convert XMLDepth to numerical depth.");
        return -1;
    }
}
```

@rboulton-BH @EKAdebo is a return of `-1` for a failed depth suitable here?

File: XML_Engine/Query/Depth.cs [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+namespace BH.Engine.Adapters.XML
+{
+    public static partial class Query
+    {
+        /***************************************************/
+        /**** Public Methods                            ****/
+        /***************************************************/
+
+        [Description("Determines whether a collection of Environment Panels representing a single space contains any panels which are externally facing")]
```
Comment (FraserGreenroyd) [inline XML_Engine/Query/Depth.cs:46]: 👀 are you sure?

File: XML_Engine/Query/Depth.cs [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        [Input("elementsAsSpace", "A collection of Environment Panels that represent closed spaced")]
+        [Input("elementsAsSpaces", "A nested collection of Environment Panels which represent all spaces in the model")]
+        [Output("isExternal", "True if the space has at least one externally facing Panel")]
```
Comment (FraserGreenroyd) [inline XML_Engine/Query/Depth.cs:49]: I suspect this might be a copy-paste that wasn't updated 😉

File: XML_Engine/XML_Engine.csproj [hunk 3]
```diff
@@ -46,6 +46,9 @@
       <Private>False</Private>
       <SpecificVersion>False</SpecificVersion>
     </Reference>
+    <Reference Include="BHoM_Engine">
+      <HintPath>..\..\..\ProgramData\BHoM\Assemblies\BHoM_Engine.dll</HintPath>
```
Comment (FraserGreenroyd) [inline XML_Engine/XML_Engine.csproj:50]:
```suggestion
      <HintPath>C:\ProgramData\BHoM\Assemblies\BHoM_Engine.dll</HintPath>
```

Comment (FraserGreenroyd) [reply XML_Engine/XML_Engine.csproj:50]: Will need further compliance on specific version which the bot will pick up for us.

**Review by FraserGreenroyd — CHANGES_REQUESTED** [review]
See comments @CKBoulter @jamesramsden-bh

File: XML_Engine/Query/Depth.cs [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+namespace BH.Engine.Adapters.XML
+{
+    public static partial class Query
+    {
+        /***************************************************/
+        /**** Public Methods                            ****/
+        /***************************************************/
+
+        [Description("Returns the 'depth' property of a markup object as a double. If the value cannot be converted to a double (if the value is null or blank for example) then an error will be returned.")]
```
Comment (FraserGreenroyd) [inline XML_Engine/Query/Depth.cs:46]:
```suggestion
        [Description("Returns the 'depth' property of a markup object as a double. If the value cannot be converted to a double (if the value is null or blank for example) then an error will be provided and -1 will be returned.")]
```

File: XML_Engine/Query/Depth.cs [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+        [Input("markup", "A markup object which contains the 'depth' property.")]
+        [Output("Depth", "The depth value as a double, or -1 if the value could not be converted to a double.")]
```
Comment (FraserGreenroyd) [inline XML_Engine/Query/Depth.cs:48]:
```suggestion
        [Output("depth", "The depth value as a double, or -1 if the value could not be converted to a double.")]
```

**Review by FraserGreenroyd — CHANGES_REQUESTED** [review]

**Comment by FraserGreenroyd** [comment]
@BHoMBot check compliance

**Comment by CKBoulter** [comment]
@BHoMBot fix project file ref. 9695000940

**Comment by FraserGreenroyd** [comment]
@BHoMBot fix project file ref. 9695000940

**Comment by FraserGreenroyd** [comment]
@BHoMBot check compliance

**Comment by FraserGreenroyd** [comment]
@BHoMBot check required

**Comment by FraserGreenroyd** [comment]
@BHoMBot check null-handling

**Comment by FraserGreenroyd** [comment]
@BHoMBot check compliance
@BHoMBot check required

**Review by FraserGreenroyd — APPROVED** [review]
Functionality remains the same following @EKAdebo review, new method looks good and works well. Happy to merge to alpha.

**Comment by FraserGreenroyd** [comment]
@BHoMBot check ready-to-merge

**Pull request merged**